### DISCUSSION TITLE
Different behaviour if running subscribe_data_change and subscribe_alarms_and_conditions in one process or different processes

### DISCUSSION THREAD

**Discussion by ts4iot** [FreeOpcUa/opcua-asyncio#1231] [Q&A]
I'm capturing data from a Siemens WinCC server. 
The alarm messages from the WinCC server are captured by **_subscribe_alarms_and_conditions_**. 
Some status words of components are captured by **_subscribe_data_change_**.

Both handlers are writing the captured data to different asyncio.Queue(s).

If I  run the data capturing of **_subscribe_alarms_and_conditions_** and **_subscribe_data_change_** in two separat procesess, the data capturing runs fine and looks well synchronized.

If I run both in one process it looks also fine at the beginning but than step by step the subscription with the higher load becomes dominant and after some time I got no information from the subscription with the lower load.

Any idea if this behaviour could be caused by the opcua-asyncio library?

**Comment by AndreasHeine** [comment]
you need to ensure your queue consumers to be at equal priority so both will get the chance to process a item!

**Comment by oroulet** [comment]
if you are overloading your CPU or not releasing the loop often enough you can get whatever strange issues. And yes what you describe is typical

**Comment by ts4iot** [comment]
My consuming code look like this. The send methods are transmitting data to an iot platform. How to check if I have enough free time in the main loop. CPU util is at 40%. 20% with only one subscription.

```
	while g_running:
		if(not g_eventqueue.empty()):
			g_logger.info("Event queue size start: %d" % g_eventqueue.qsize())	
			while(not g_eventqueue.empty()):
				send_event(await g_eventqueue.get())	
		if(not g_statusqueue.empty()):
			g_logger.info("Status queue size start: %d" % g_statusqueue.qsize())		
			while(not g_statusqueue.empty()):
				send_state(await g_statusqueue.get())	
		await g_plc.check_connection()
		await asyncio.sleep(0.1)
```

The event queue receives about 100 messages / second.
The status queue receives about 25 messages / second.

**Reply by AndreasHeine** [reply]
currently you process all items in one queue and after that you process all items of the second one... and with the "while queue not empty", you will never have equal priority, so as long as one queue is not empty it will not switch to the other... you should consider using one queue instead with two different itemtypes in it or process one item (if not queue not empty) from each queue in a whileloop so it will process both (one item from both) in one iteration.

**Reply by AndreasHeine** [reply]
the limit of python 3.10 asyncio is beyond 600+ msg/sec (on a i7 9750H on Win10 -> WindowsSelectorEventLoopPolicy... so on linux with ProactorEventLoop its faster) but you need streamlined code which does not block at all!

**Comment by ts4iot** [comment]
I also made some tests without sending the data to the iot platform. But the behaviour is the same.

**Comment by ts4iot** [comment]
Thank you for the help. After disabling the transmission to the iot platform and decreasing the **_await asyncio.sleep(0.1)_** to **_await asyncio.sleep(0.01)_** brings the expected result. This is a clear indication that the data transmission needs to much time and blocks the opcua processing. Currently i use paho-mqtt for the data transmission. I will now switch to asyncio-mqtt.

**Reply by AndreasHeine** [reply]
see:
https://github.com/AndreasHeine/opcua-sub-to-mqtt/blob/c9816fcf1ea0c91a22a06f680263c6df59640881/opcua_client_mqtt_publisher.py#L272
